src/openzeppelin/token/erc1155.cairo

```diff
@@ -0,0 +1,327 @@
+    }
+
+    #[view]
+    fn uri(id: u256) -> felt252 {
+        ERC1155::uri(id)
+    }
+
+    #[external]
+    fn set_approval_for_all(operator: ContractAddress, approved: bool) {
+        ERC1155::set_approval_for_all(operator, approved)
+    }
+
+    #[external]
+    fn safe_transfer_from(
+        from: ContractAddress, to: ContractAddress, id: u256, amount: u256, data: Array<felt252>
+    ) {
+        ERC1155::safe_transfer_from(from, to, id, amount, data)
+    }
+
+    #[external]
+    fn safe_batch_transfer_from(
+        from: ContractAddress,
+        to: ContractAddress,
+        ids: Array<u256>,
+        amounts: Array<u256>,
+        data: Array<felt252>
+    ) {
+        ERC1155::safe_batch_transfer_from(from, to, ids, amounts, data)
+    }
+
+    #[internal]
+    fn initializer(uri: felt252) {
+        _set_uri(uri);
+        erc165::ERC165Contract::register_interface(erc1155::IERC1155_ID);
+        erc165::ERC165Contract::register_interface(erc1155::IERC1155_METADATA_ID);
+    }
+
+    #[internal]
+    fn _set_approval_for_all(owner: ContractAddress, operator: ContractAddress, approved: bool) {
+        assert(owner != operator, 'ERC1155: self approval');
+        _operator_approvals::write((owner, operator), approved);
+        ApprovalForAll(owner, operator, approved);
+    }
+
+    #[internal]
+    fn _safe_transfer_from(
+        from: ContractAddress, to: ContractAddress, id: u256, amount: u256, data: Array<felt252>
+    ) {
+        let operator: ContractAddress = get_caller_address();
+        assert(!to.is_zero(), 'ERC1155: invalid receiver');
+
+        _balances::write((id, from), _balances::read((id, from)) - amount);
+        _balances::write((id, to), _balances::read((id, to)) + amount);
+        TransferSingle(operator, from, to, id, amount);
+
+        _do_safe_transfer_acceptance_check(operator, from, to, id, amount, data);
+    }
+
+    #[internal]
+    fn _safe_batch_transfer_from(
+        from: ContractAddress,
+        to: ContractAddress,
+        ids: Array<u256>,
+        amounts: Array<u256>,
+        data: Array<felt252>
+    ) {}
+
+    #[internal]
+    fn _set_uri(uri: felt252) {
+        _uri::write(uri)
+    }
+
+    #[private]
+    fn _asSingletonArray(value: u256) -> Array<u256> {
+        let mut array = ArrayTrait::new();
+        array.append(value);
+        array
+    }
+
+    #[private]
+    fn _balance_of_batch_iter(
+        mut accounts: Array<ContractAddress>, mut ids: Array<u256>, mut res: Array<u256>
+    ) -> Array<u256> {
+        match accounts.pop_front() {
+            Option::Some(account) => {
+                let id = ids.pop_front().expect('ERC1155 invalid array length');
+                res.append(ERC1155::balance_of(account, id));
+                _balance_of_batch_iter(accounts, ids, res)
+            },
+            Option::None(_) => {
+                assert(ids.is_empty(), 'ERC1155 invalid array length');
+                res
+            }
+        }
+    }
```

Comment:
This will be replaced by a loop ([see this commit](https://github.com/starkware-libs/cairo/commit/06827b725060538db66124b8fbd08465783a45d2)) when the next alpha is available.